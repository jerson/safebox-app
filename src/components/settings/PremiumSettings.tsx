import React, { useState } from 'react';
import {
  StyleSheet,
  ViewStyle,
  View,
  StyleProp,
  Alert,
  TextStyle,
  Platform
} from 'react-native';
import Colors from '../../modules/constants/Colors';
import SplitText from '../ui/SplitText';
import Button from '../ui/Button';
import Text from '../ui/Text';
import * as RNIap from 'react-native-iap';
import Log from '../../modules/log/Log';
import Client from '../../services/Client';
import Strings from '../../modules/format/Strings';
import { HasProductRequest, BuyProductRequest } from '../../proto/services_pb';
import AlertMessage from '../ui/AlertMessage';

const styles = StyleSheet.create({
  container: {
    backgroundColor: Colors.grey1
  } as ViewStyle,
  content: {
    marginTop: 10,
    flexDirection: 'row',
    alignItems: 'center'
  } as ViewStyle,
  info: {
    flex: 1,
    paddingRight: 10
  } as ViewStyle,
  description: {
    fontSize: 12,
    color: Colors.grey5
  } as TextStyle,
  button: {
    minWidth: 160
  } as ViewStyle,
  splitText: {
    marginBottom: 5
  } as ViewStyle
});

export interface PremiumSettingsProps {
  style?: StyleProp<ViewStyle>;
}

const TAG = '[PremiumSettings]';
function PremiumSettings({ style }: PremiumSettingsProps) {
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState('');

  const hasProduct = async (productId: string) => {
    try {
      const request = new HasProductRequest();
      request.setSlug(productId);
      const response = await Client.hasProduct(request);
      return response.getPurchased();
    } catch (e) {
      return false;
    }
  };

  const sendPurchase = async (purchase: RNIap.ProductPurchase) => {
    const payload = purchase.purchaseToken || purchase.transactionReceipt || '';
    try {
      const request = new BuyProductRequest();
      request.setSlug(purchase.productId);
      request.setPayload(payload);
      request.setType(Platform.OS);
      await Client.buyProduct(request);
      return true;
    } catch (e) {
      Log.warn(TAG, 'sendPurchase', e);
      return false;
    }
  };

  const restore = async () => {
    setIsLoading(true);
    let totalRestored = 0;
    try {
      const purchases = await RNIap.getAvailablePurchases();

      for (const purchase of purchases) {
        const hasProductValidation = await hasProduct(purchase.productId);
        if (!hasProductValidation) {
          const isValid = await sendPurchase(purchase);
          isValid && totalRestored++;
        }
      }
    } catch (e) {
      const message = Strings.getError(e);
      setError(message);
    }
    setIsLoading(false);

    if (totalRestored > 0) {
      Alert.alert(`${totalRestored} products restored`);
    }
  };

  return (
    <View style={[styles.container, style]}>
      <SplitText
        style={styles.splitText}
        title={'Premium settings'}
        type={'Default'}
      />
      {!!error && (
        <AlertMessage
          onTimeout={() => {
            setError('');
          }}
          message={error}
        />
      )}
      <View style={styles.content}>
        <View style={styles.info}>
          <Text style={styles.description}>
            If you made a purchase before and do not see it in the premium
            section, you can restore it here.
          </Text>
        </View>
        <Button
          style={styles.button}
          isLoading={isLoading}
          title={'Restore purchases'}
          typeColor={'accentDark'}
          onPress={restore}
        />
      </View>
    </View>
  );
}

export default PremiumSettings;
