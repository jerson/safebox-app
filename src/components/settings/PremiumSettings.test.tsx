import React from 'react';
import * as renderer from 'react-test-renderer';
import PremiumSettings, {PremiumSettingsProps} from './PremiumSettings';

const props: PremiumSettingsProps = {};
test('renders without crashing', () => {
  const snapshot = renderer.create(<PremiumSettings {...props} />).toJSON();
  expect(snapshot).toBeTruthy();
});

test('render just views', () => {
  const snapshot = renderer.create(<PremiumSettings {...props} />).toJSON();
  expect(snapshot).toMatchSnapshot();
});
