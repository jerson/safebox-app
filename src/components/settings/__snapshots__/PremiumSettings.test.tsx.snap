// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render just views 1`] = `
<_class
  style={
    Array [
      Object {
        "backgroundColor": "#fafdff",
      },
      undefined,
    ]
  }
>
  <_class
    style={
      Array [
        Object {
          "alignItems": "center",
        },
        Object {
          "marginBottom": 5,
        },
      ]
    }
  >
    <_class
      style={
        Array [
          Object {
            "backgroundColor": "#9066f5",
            "height": 1,
            "left": 0,
            "opacity": 0.5,
            "position": "absolute",
            "right": 0,
            "top": 12,
          },
          Object {
            "backgroundColor": "#c2d0e7",
          },
          false,
        ]
      }
    />
    <_class
      style={
        Array [
          Object {
            "backgroundColor": "#4531cd",
            "borderRadius": 10,
            "paddingBottom": 4,
            "paddingLeft": 8,
            "paddingRight": 8,
            "paddingTop": 6,
          },
          Object {
            "backgroundColor": "#6781a6",
          },
          false,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Nunito",
              "fontSize": undefined,
              "fontStyle": "normal",
              "fontWeight": "400",
            },
            Object {},
            Object {
              "color": "#FFFFFF",
              "fontSize": 13,
              "lineHeight": 14,
              "minWidth": 20,
              "textAlign": "center",
            },
          ]
        }
      >
        Premium settings
      </Text>
    </_class>
  </_class>
  <_class
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "marginTop": 10,
      }
    }
  >
    <_class
      style={
        Object {
          "flex": 1,
          "paddingRight": 10,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Nunito",
              "fontSize": undefined,
              "fontStyle": "normal",
              "fontWeight": "400",
            },
            Object {},
            Object {
              "color": "#65748b",
              "fontSize": 12,
            },
          ]
        }
      >
        If you made a purchase before and do not see it in the premium section, you can restore it here.
      </Text>
    </_class>
    <_class
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#6781a6",
          "borderColor": "#6781a6",
          "borderRadius": 10,
          "flexDirection": "row",
          "justifyContent": "center",
          "margin": 0,
          "minWidth": 160,
          "opacity": 1,
          "padding": 12,
          "zIndex": 1,
        }
      }
      testID="button_touchable"
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Nunito",
              "fontSize": undefined,
              "fontStyle": "normal",
              "fontWeight": "400",
            },
            Object {},
            Array [
              Object {
                "color": "#FFFFFF",
                "fontSize": 14,
                "textAlign": "center",
              },
              Object {
                "color": "#FFFFFF",
              },
              undefined,
            ],
          ]
        }
      >
        Restore purchases
      </Text>
    </_class>
  </_class>
</_class>
`;
